Optimizing `raster.icrop`

vschaffn opened the pull request:
Resolves #659 

## Description
- Changes order of `icrop` bounding box to match with Rasterio.BoundingBox.
- Adds a parameter `distance_unit: Literal["georeferenced", "pixel"] = "georeferenced"` in `georeferencing._crop` to match with the use of `icrop`.

## Tests
- Updates crop tests.

## Doc
-  Updates crop section in `raster_class.md`.

adebardo commented:
Are we sure that the DEM is not completely load before crop ?

rhugonnet commented on geoutils/raster/georeferencing.py:
```diff
@@ -173,7 +173,7 @@ def _outside_image(
 
     if not index:
-        xi, xj = _xy2ij(xi, yj, transform=transform, area_or_point=area_or_point)
+        xi, yj = _xy2ij(xi, yj, transform=transform, area_or_point=area_or_point)
```
Good catch! :open_mouth:

rhugonnet commented on geoutils/raster/raster.py:
```diff
@@ -2464,15 +2480,29 @@ def icrop(
         if len(bbox) != 4:
             raise ValueError("bbox must be a list or tuple of four integers: (xmin, ymin, xmax, ymax).")
 
-        # Convert pixel coordinates to georeferenced coordinates using the transform
```
The argument description above was ambiguous for me:

>         :param bbox: Pixel-based bounding box as (xmin, ymin, xmax, ymax) in pixel coordinates.

Maybe similar language as in the function description "Bounding box based on indices of the raster array".

We could also add a note in the general description to clarify which this is strikingly different, something like: "Note that, by convention, a raster array has the Y axis as first in the array, and that this Y axis is also flipped (first index corresponds to the highest-value coordinate)."

rhugonnet commented on geoutils/raster/raster.py:
```diff
@@ -2464,15 +2480,29 @@ def icrop(
+        # Crop the raster data based on the given bounding box (bbox), limit the range to the borders of the raster
+        crop_img = self.data[..., bbox[0] : min(self.height, bbox[2]), bbox[1] : min(bbox[3], self.width)]
+
+        # Create a new affine transform that corresponds to the cropped area
```
Took me a little while to understand why this transform looked like a translation!
When we crop within the extent, the new array size determines the new height/width, so it is only a translation for the affine transform!
In this case, let's use our tested `_translate` function to make it easier to read/debug for future us :wink::
https://github.com/GlacioHack/geoutils/blob/b4da49b8e5ff26f3120085c01cbaedbc1ee1b10b/geoutils/raster/geotransformations.py#L533

rhugonnet commented on tests/test_raster/test_raster.py:
```diff
@@ -1818,28 +1818,33 @@ def test_icrop(self, example: str, bbox: list[int] | tuple[int, ...]) -> None:
+        raster_icropped = raster.icrop(bbox=bbox)
 
         # Compute the expected affine transform for the cropped raster
         transformed_cropped = rio.transform.Affine(
```
Can use `_translate` here too.

rhugonnet reviewed (approved):
Perfect, thanks!

rhugonnet commented on geoutils/raster/raster.py:
```diff
@@ -2464,15 +2480,29 @@ def icrop(
-        bottom_left = self.transform * (bbox[0], self.height - bbox[1])
-        top_right = self.transform * (bbox[2], self.height - bbox[3])
+        # Crop the raster data based on the given bounding box (bbox), limit the range to the borders of the raster
+        crop_img = self.data[..., bbox[0] : min(self.height, bbox[2]), bbox[1] : min(bbox[3], self.width)]
```
Ah sorry I didn't notice this line and @adebardo's comment, good catch.
Indeed, this will load the entire array, then subset it. To circumvent this, we need to use `crop()` as before.

rhugonnet replied:
We should add a test checking if `Raster.is_loaded` is still `False` after calling `icrop()`.

vschaffn commented:
@rhugonnet @adebardo thanks for your feedback !
I have corrected the `Raster.icrop` method : 
- If the raster is not loaded, I use `_load_rio` with a window to just load the cropped raster
- If the raster is already loaded I kept the same way to do
- I replaced the transform translation by `Raster.translate`

I also fixed a sign error in `_translate`, and added some more tests

The loading check can't be set up in the tests because the raster is always loaded after `icrop`

rhugonnet commented:
Thanks!

I'm not sure I understand the logic of the part of the PR regarding removing the dependency to `crop()`? To me that's a huge benefit. It allows to have a fully consistent behaviour, with lots of tests already written in `crop()`, without having to duplicate the same code (what is done here in the end, by reading with `rio.Windows`).
And it doesn't affect performance: converting 4 indices to 4 coordinates is instantaneous.

vschaffn commented:
@adebardo @rhugonnet @adehecq I have changed the implementation of the `icrop` as discussed on 28/03/2025. Does this work for you? 😃

rhugonnet commented:
Perfect for me! Great idea to make it work through a `distance_unit` argument for consistency with other functions :wink:

adebardo commented:
Thanks for the work @vschaffn ! :) and @rhugonnet for the review :) I merge
